Proposal to recover broker-info.subm file from running broker (#144)
submariner-io/enhancements: subctl/issue-89-reconstructing-broker-info.md

# Reconstruct broker-info.subm file from the running Broker

Related Issue:
[Option for recovering the broker-info.subm file from a running broker cluster](https://github.com/submariner-io/enhancements/issues/143)

## Summary

`broker-info.subm` file is created when a broker is installed. The file contains information critical to have a functional environment. 
There is a possibility that the file could be deleted or lost for one or the other reason. In such cases, the environment becomes redundant
and no new clusters could be joined.

This proposal describes a way to reconstruct the `broker-info.subm` file from the running broker.

## Proposal

A new `subctl` command will be added that would fetch needed data from the broker and write it to `broker-info.subm` file. The file 
would be Base64 encoded.

## Design details

> * subctl command name: `reconstruct-brokerinfo`.

Reviewer: ```suggestion
* subctl command name: `recover-broker-info`.
```

Author: Well, as per definitions in this [wikidiff](https://wikidiff.com/reconstruct/recover), we are constructing a **new file** with needed info. We are not getting the **same file** which was lost.

Author: Done

* Flags for the command:
  * `namespace` to specify the namespace in which broker is installed.
  * `ipsec-psk-from` to specify the file to read PSK secret from.
  * All flags that `restConfigProducer` provides.
* Data to be fetched from running broker:
  * Broker URL
  * Client token
  * IPSec PSK
  * Components installed i.e. Submariner and Service discovery
  * whether service discovery component is installed
  * default custom domains

The last three information will be retrieved from the broker `Spec`, Broker URL from the user provided kubeconfig, client token from the 
Broker admin service account `submariner-k8s-broker-admin` and IPSec PSK with the following logic:

```
if file specified via `ipsec-psk-from` flag by the user
  read the secret from the specified file
else if submariner is installed in the cluster
  read the secret from submariner spec

> else

Reviewer 2: There’s a missing case here: if the user didn’t provide a file, and Submariner isn’t installed (or isn’t accessible using the available credentials), but the broker has one or more connected clusters, `subctl` should fail with an error — a `broker-info.subm` with a new PSK would not allow newly-connected clusters to connect to existing clusters.

Author: Done.

# Reconstruct broker-info.subm file from the running Broker

Related Issue:
[Option for recovering the broker-info.subm file from a running broker cluster](https://github.com/submariner-io/enhancements/issues/143)

## Summary

`broker-info.subm` file is created when a broker is installed. The file contains information critical to have a functional environment.
There is a possibility that the file could be deleted or lost for one or the other reason. In such cases, no new clusters could be joined.

This proposal describes a way to reconstruct the `broker-info.subm` file from the running broker.

## Proposal

A new `subctl` command will be added that would fetch needed data from the broker and write it to `broker-info.subm` file. The file
would be Base64 encoded.

## Design details

* subctl command name: `reconstruct-brokerinfo`.
* Flags for the command:

>   * `submconfig` to specify the cluster in which Submariner is installed

Reviewer 3: Not sure I understand this. What exactly is this `submconfig`? and how is it related to `kubeconfig`? 
Does `submconfig` mean - SubmarinerConfig which is only applicable in OCM/ACM deployments?

Author: @sridhargaddam its a kubeconfig of a cluster which has Submariner installed.

Author: `submconfig` is an apt name for this flag. And anyways this is very subctl specific and only all Submariner users are OCM users.

>   * `submconfig` to specify kubeconfig of the cluster in which Submariner is installed

Author: rename this name. Suggestions?

Author: Address the case where the cluster has both broker and Submariner installed so no need to provide 2 kubeconfigs.

Author: Done.

# Recover broker-info.subm file from the running Broker

Related Issue:
[Option for recovering the broker-info.subm file from a running broker cluster](https://github.com/submariner-io/enhancements/issues/143)

## Summary

`broker-info.subm` file is created when a broker is installed. The file contains information critical to have a functional environment.
There is a possibility that the file could be deleted or lost for one or the other reason. In such cases, no new clusters could be joined.

This proposal describes a way to recover the `broker-info.subm` file from the running broker.

## Proposal

A new `subctl` command will be added that would fetch needed data from the broker and write it to `broker-info.subm` file. The file
would be Base64 encoded.

## Design details

* subctl command name: `recover-brokerinfo`
* Flags for the command:

>   * `submconfig` to specify kubeconfig of the cluster in which Submariner is installed

Reviewer 3: As mentioned during the call, I'm personally not convinced that `submConfig` is an apt name as we use it in ACM/OCM. Please come up with a better name.

Reviewer: I think it should be the normal `--context`. As discussed,

- First check If the `Submariner` (or `ServiceDiscovery`) resource exists, if so proceed with recovery
- If not then check if the broker is installed (see [findBrokerNamespace](https://github.com/submariner-io/subctl/blob/3d44295393e4ab4ca35976030e93aa194a3e2c3d/pkg/uninstall/all.go#L324))
- If broker installed, then check if it's in use (ie check if there are any `Endpoints` in the broker NS)
- If not in use then generate a new PSK
- If in use then exit and inform the user they must provide the kubeconfig of a joined cluster

Reviewer 2: That only works if we can recover everything using _only_ a `Submariner` resource; otherwise we need to be able to specify two contexts (one to access a broker, the other to access a joined cluster).

Reviewer: The `Submariner` resource contains the connectivity info to access the broker, if needed.

Author: So, the original kubeconfig provided should be of Broker cluster. And if Submariner is not installed on that cluster, and there is a joined cluster, provide kubeconfig of the cluster that has Submariner installed. Right?

Reviewer: As outlined in the steps I posted above, assuming we need the `Submariner` resource from a previously joined cluster in order to recover the broker info, then we expect the user to provide the kubeconfig of a joined cluster. If not and it's the broker cluster then we can try to recover iff there are no joined clusters.

Author: Client token is stored only on `broker`. So access to broker (ad hence its kubeconfig) is necessary. Unless we get broker restconfig from Submariner/service discovery like [gather does](https://github.com/submariner-io/subctl/blob/devel/internal/gather/gather.go#L183)

Reviewer: > Unless we get broker restconfig from Submariner/service discovery like [gather does](https://github.com/submariner-io/subctl/blob/devel/internal/gather/gather.go#L183)

Right - that's what I said above, also like the pods do.

Author: Turns out serviceaccounts are not accessible from the Submariner cluster.
`error getting client token ServiceAccount submariner-k8s-broker-admin get failed: serviceaccounts "submariner-k8s-broker-admin" is forbidden: cannot get resource "serviceaccounts" in API group "" in the namespace "submariner-k8s-broker"`

Reviewer: So the broker credentials in the `Submariner` object used by the pods to sync with the broker NS are associated with the "_submariner-k8s-broker-cluster_" `Role` defined [here](https://github.com/submariner-io/subctl/blob/98f65ae7e95e0f5ebe695545acd03d4e815a2cb5/pkg/broker/rbac.go#L82).  So the `Role` doesn’t allow access to SA’s although it does allow access to `Secrets`. You need the `Secret` associated with the “_submariner-k8s-broker-admin_” SA but I guess the only way to get the `Secret` name is [via the SA](https://github.com/submariner-io/subctl/blob/98f65ae7e95e0f5ebe695545acd03d4e815a2cb5/internal/rbac/rbac.go#L36). 

Unless we want to grant the "_submariner-k8s-broker-cluster_" `Role` access to SA's then we do need the user to pass the broker kubeconfig but I don't see this as a compelling reason to expand the privileges. The kubeconfig of a joined cluster would be needed as well except if there's no joined clusters.

Reviewer: Back to the original comment, it seems we've deprecated the passing of kubeconfigs, so how about `--context` for the joined cluster and `--broker-context`.

Reviewer 2: We haven’t deprecated the passing of kubeconfigs, we’ve made them consistent; so `--kubeconfig` and `--context` for the joined cluster, `--brokerconfig` and `--brokercontext` for the broker (following existing practice).

  * `namespace` to specify the namespace in which broker is installed
  * `ipsec-psk-from` to specify the file to read PSK secret from
  * All flags that `restConfigProducer` provides
* Data to be fetched from running broker:
  * Broker URL
  * Client token
  * IPSec PSK
  * Components installed i.e. Submariner and Service discovery
  * whether service discovery component is installed
  * default custom domains

The last three information will be retrieved from the broker `Spec`, Broker URL from the user provided kubeconfig, client token from the
Broker admin service account `submariner-k8s-broker-admin` and IPSec PSK with the following logic:

> ```mermaid

Author: Review this under `rich diff` format (select from the top right button) to see the flowchart

# Recover broker-info.subm file from the running Broker

Related Issue:
[Option for recovering the broker-info.subm file from a running broker cluster](https://github.com/submariner-io/enhancements/issues/143)

## Summary

`broker-info.subm` file is created when a broker is installed. The file contains information critical to have a functional environment.
There is a possibility that the file could be deleted or lost for one or the other reason. In such cases, no new clusters could be joined.

This proposal describes a way to recover the `broker-info.subm` file from the running broker.

## Proposal

A new `subctl` command will be added that would fetch needed data from the broker and write it to `broker-info.subm` file. The file
would be Base64 encoded.

## Design details

> * subctl command name: `recover-brokerinfo`

Reviewer 3: Can you please highlight/show the mandatory flags and the optional flags?

I guess the mandatory flags are `brokerconfig` and all other flags are optional.

Reviewer: Neither necessarily has to be mandatory. `--kubeconfig` points to a possible `Submariner` cluster and `--brokerconfig` points to the possible broker cluster. If only `--kubeconfig` is specified and it also has the broker installed then we can proceed w/o needing `--brokerconfig`. Similarly, if only `--brokerconfig` is specified and it also has `Submariner` installed then we can proceed.

Reviewer 3: Right, it was not clear in the description. But anyways, if this is how its implemented, its good.

* Flags for the command:
  * `brokerconfig` to specify kubeconfig of the cluster in which Broker is installed
  * `brokernamespace` to specify the namespace in which broker is installed
  * `ipsec-psk-from` to specify the file to read PSK secret from
  * All flags that `restConfigProducer` provides
    * `kubeconfig` to specify kubeconfig of the cluster in which Submariner is installed
* Data to be fetched from running broker:
  * Broker URL
  * Client token
  * IPSec PSK
  * Components installed i.e. Submariner and Service discovery
  * whether service discovery component is installed
  * default custom domains

> The last three information will be retrieved from the broker `Spec`, Broker URL from the user provided broker kubeconfig, client token 

Reviewer 2: ```suggestion
The last three information will be retrieved from the broker `Spec`, Broker URL from the user provided broker kubeconfig, client token
```

from the Broker admin service account `submariner-k8s-broker-admin` and IPSec PSK with the following logic:

```mermaid
flowchart TD
  A{Submariner config provided via `--kubeconfig` flag?} -- Yes ---> D
  A -- No ---> B
  B[Broker kubeconfig provided via `--brokerconfig` flag] ---> C{Broker cluster has Submariner installed?}
  C -- Yes ---> D[Get PSK secret from submariner.Spec]
  C -- No ---> E{"Any cluster joined to Broker (Figure out from the endpoints)?"}
  E -- Yes ---> F[Exit]
  E -- No ---> G{PSK file specified via `--ipsec-psk-from` flag?}
  G -- Yes ---> H[Use secret from the file]
  G -- No --> I[Generate new secret]
```

> If Submariner is not installed in the cluster specified by `--config` flag, exit the execution and throw an error. This addresses

Reviewer 2: ```suggestion
If Submariner is not installed in the cluster specified by the `--kubeconfig` flag, end the execution and throw an error. This addresses
```

the scenario where there is a cluster with Submariner in the environment but an incorrect config was provided via the flag. In such
scenario, if the execution is not stopped, a different/new PSK secret would be populated which would break the environment.

### Usage

> The new command can be used as follows (assuming Submariner is installed in cluster2:

Reviewer 3: ```suggestion
The new command can be used as follows (assuming Submariner is installed in cluster2):
```

Reviewer 3: Small nit, if you happen to re-spin, otherwise its fine.

> `subctl recover-brokerinfo --brokerkubeconfig output/kubeconfigs/kind-config-cluster1 --kubeconfig output/kubeconfigs/kind-config-cluster2`

Reviewer 2: ```suggestion
`subctl recover-brokerinfo --brokerconfig output/kubeconfigs/kind-config-cluster1 --kubeconfig output/kubeconfigs/kind-config-cluster2`
```

> `subctl recover-brokerinfo --brokerconfig output/kubeconfigs/kind-config-cluster1 --kubeconfig output/kubeconfigs/kind-config-cluster2`

Reviewer 3: Please add another example, where the cluster1 is used both as Broker and Submariner cluster.
